### ISSUE TITLE
Emit: unstable for emit to specific user

### ISSUE THREAD

**Issue by kvdmolen** [deployd/deployd#533]
There seems to be a recurring issue with emit:

I am emitting data to one specific user (for messaging service) using:
```javascript
emit(dpd.users, {id:..}, "action", this)
```
which works, but only now and then! Very arbitrary when it does work and when it doesn't

Also it seems that after some minutes of inactivity the connection is lost, and then the emit stops working. A refresh does give me the new messages so that is not the problem (dpd 0.8.0).

Thanks for the help.

**Comment by andreialecu** [comment]
I think I know why this is happening. Could you, as a test, refresh the page in the browser *after* logging in once, then checking if you still have the issue? 

Implementation notes:
I suspect that the socket.io connection is not properly attached to an user if you are using the REST api to log in because the sid cookie was not present when the socket.io connection was established and there is no way to associate the two.

**Comment by andreialecu** [comment]
I have been thinking about this for a while, as I discovered it while auditing the code. I think we need to force a socket.io disconnect/reconnect after a successful login.

**Comment by ericfong** [comment]
Will angular-dpd have the same issue?

**Comment by andreialecu** [comment]
Yes - because it is socket.io session related and angular-dpd doesn't do anything special in that regard.

However, simply disconnecting and reconnecting socket.io immediately after successful `dpd.users.login()` should probably be enough to resync the connection. It is possible to call `.disconnect()` then `.reconnect()` on the underlying socket.io `io` manager class.

I haven't personally ran into this problem yet because I'm not emitting to certain users yet, only to all. But I have noticed that it could occur just by auditing the code in `sessions.js`.

We need @kvdmolen to confirm whether this workaround first.

**Comment by andreialecu** [comment]
I think I have a fix for this at #540

**Comment by kvdmolen** [comment]
Thanks for all the comments! I've used Pusher.com in the meantime, but I'll check the suggestions posted here a.s.a.p. and get back with results.

**Comment by xroussel** [comment]
I have the same problem using deployd + angular-dpd.
I try to create a simple chat application where an user can send messages to a specific user.
I can connect my users and transmit a few messages but after 2/3 messages (very random),  the notifications are not sent anymore.

Server :
```javascript
    emit(dpd.users, {id:"..."}, 'newMessage', {...}) ;
```
Client:
```javascript
    dpd.users.exec('login', {username:"...",password:"..."})
    .success(function(re) {
	    ...
	    // Receive messages
	    dpd.on($rootScope, 'newMessage', function(message) {
		    console.log(message) ;
	    });

        ...

	    // Send message
	    dpd.messages.post({...}, function(message) {
		    console.log(message) ;
	    }) ;
	    ...
    });
```
I used a workaround in angular-dpd to be able to establish the socket after user connection (just create dpdSocket after login) but the problem persist after the connection.
It seems that the sockets stored by deployd in ctx.session.store.socketIndex are removed when a user send data using the $http service, if the same user continually send message to the other, the socket is not destroyed and the messages are sent properly.
I also tried to apply the patch #540 and use the last version available but that didn't resolved the bug for me.

**Comment by ericfong** [comment]
Any proxy ? load balancer or nginx in front of dpd? 

I got some problem with some load balancer software before.

**Comment by xroussel** [comment]
No nothing, I have everything locally.
I did a few more tests and the bug appears only on versions >= 0.8.1 of deployd, the bug is not present on 0.8.0

**Comment by andreialecu** [comment]
Can you provide a link to a sample project showing this problem?

**Comment by xroussel** [comment]
I created a sample project to show you the bugs here :
https://github.com/xroussel/deployd-emit-issue.git

Juste create "user1" and "user2" in users db with password:"password" and test, it should work with 0.8.0 but not with >=0.8.1.
The connection bug is present in all versions (>= 0.8.0) for me.

![deployd](https://cloud.githubusercontent.com/assets/11319013/7413809/15d64d1a-ef4f-11e4-9662-4d5590647151.gif)

**Comment by ericfong** [comment]
Clone and got the same problem with

emit(dpd.users, {id:user.id}, 'newMessage', {content:msg.content});

But

emit('newMessage', this);

work

That means: lib/session.js

this.emitToAll = function()  works

and

userSessionIndex[user.id] is exists

        userSession = userSessionIndex[users.id];
        if(userSession && userSession.socket) {
          userSession.socket.emit(event, data);
          console.error(users.username, event, data);
          // OUTPUT seems the username and socket messed up
        }


Still looking into it
Not sure why, welcome hacks and PR.



On Thu, Apr 30, 2015 at 9:44 PM, Xavier Roussel <notifications@github.com>
wrote:

> I created a sample project to show you the bugs here :
> https://github.com/xroussel/deployd-emit-issue.git
>
> Juste create "user1" and "user2" in users db with password:"password" and
> test, it should work with 0.8.0 but not with >=0.8.1.
> The connection bug is present in all versions (>= 0.8.0) for me.
>
> [image: deployd]
> <https://cloud.githubusercontent.com/assets/11319013/7413809/15d64d1a-ef4f-11e4-9662-4d5590647151.gif>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/deployd/deployd/issues/533#issuecomment-97796990>.
>



-- 
Best Regards,
Eric Fong

**Comment by xroussel** [comment]
I dug a bit deeper and the bug seems to start at "ca7ac73".
In lib/session.js, the new block at line 109 :

```javascript
      if (!sess.data.anonymous && (!sess.data.lastActive || sess.data.lastActive < Date.now() - 10 * 1000)) {
        // update last active date at max once every 10 seconds
        sess.data.lastActive = Date.now();
        sess.save(function () {
          fn(null, sess);
        });
      } else {
        fn(null, sess);
      }
```

calls the function "sess.save" every 10 seconds which seems to delete all the session variables by calling "session.remove" but which does not rebuilds "session.store.socketIndex" nor "sessionIndex" properly.

Removing the line 343 of lib/session.js (last version) :
"delete session.store.socketIndex[data.id];"
solves the bug which disconnects the socket after login.

**Comment by andreialecu** [comment]
Should be fixed by #572
